packages/dds/tree/src/transaction/transaction.ts

```diff
@@ -0,0 +1,105 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IEditableForest, IForestSubscription } from "../forest";
+import { ChangeRebaser } from "../rebase";
+import { AnchorSet } from "../tree";
+
+// TODO: move this to changeset directory
+export abstract class ProgressiveEditBuilder<TChange> {
+    private readonly changes: TChange[] = [];
+    constructor(
+        private readonly changeFamily: ChangeFamily<unknown, TChange>,
+        private readonly deltaReceiver: (delta: Delta) => void,
+        private readonly anchorSet: AnchorSet) {}
+
+    /**
+     * Subclasses add editing methods which call this with their generated edits.
+     */
+    protected applyChange(change: TChange): void {
+        this.changes.push(change);
+        this.changeFamily.rebaser.rebaseAnchors(this.anchorSet, change);
```

Comment:
Once we support more anchor types (like place anchors), we may want to rebase them differently in response to local and remote edits. For example, in a text editor you might want your cursor to rebase to the left of text inserted at its location, while moving to the right of any text you insert.

Reply:
In such a case I think it would make more sense to have a different kind of anchor for that for that policy than handle local vs remote edits differently (which risks desync since the same edit is expected to be the same for everyone).

Generally, rebasing over local edits due to concurrent transactions on the same client (ex: via async transactions) should be treated the same as remote edits, however you might want to treat the anchors the app holds onto across its own edits within a transaction differently than when they are actually put into edits. The client can do this by creating the anchor with the desired semantics when creating it, and might convert it if using it in the actual edit.